### DISCUSSION TITLE
Allow to cancel a pending callback

### DISCUSSION THREAD

**Discussion by na2axl** [cycfi/elements#361] [Ideas]
Hello there,

I'm currently using elements for a project (the studio app of my library [here](https://amplitudeaudiosdk.com)), and I have very happy times with it.

Now I'm facing a challenge that may be easily fixable from my point of view. Basically I want to cancel a pending callback on the view's timer, so if it was not called before the timeout when cancelled, it is not called at all.

An example may be:

```cpp
_canceller = _view.post(2s, []() { /* do something... */ });

// Something happened elsewhere and now the timed callback should be cancelled
_canceller.cancel();
```

Digging in the elements code I've seen the `post` method may return the `timer` object to achieve that. Or if that object may lead to wrong usage from the developer, we can instead add a wrapper around it that only provide a `cancel` method. What do you think?

https://github.com/cycfi/elements/blob/f54953d256e2a218fc8bf1e9926d7cbe552f63ed/lib/include/elements/view.hpp#L329-L341

**Comment by djowel** [comment]
> I'm currently using elements for a project (the studio app of my library [here](https://amplitudeaudiosdk.com)), and I have very happy times with it.

Before anything else, please add your library here: https://github.com/cycfi/elements/discussions/352

Your library looks great! Very nice docs you have there. Are there screenshots, BTW?

**Reply by na2axl** [reply]
Hey @djowel, thank you for the warm comments 🙏🏾 

ATM I have only 2 windows, only the Welcome window is fully functional:

*Welcome Window*:

<img width="912" alt="image" src="https://github.com/cycfi/elements/assets/14062034/eaf96bb3-53cb-469c-bd16-29406b88cd27">

<img width="912" alt="image" src="https://github.com/cycfi/elements/assets/14062034/bd6b5f92-e07a-45f1-84f9-12dfd24b5dfb">

<img width="912" alt="image" src="https://github.com/cycfi/elements/assets/14062034/ac94e352-1328-4212-bea8-f15d849b918f">

*Studio Window*:

<img width="1517" alt="image" src="https://github.com/cycfi/elements/assets/14062034/ce3ae109-0f74-4b68-860e-8707d02ea642">

<img width="1517" alt="image" src="https://github.com/cycfi/elements/assets/14062034/af3adb7b-8f90-4ad8-8703-fe32d6514e14">

As the studio app is not yet available (and is closed source), I don't know if it's right to add it to #352 now. But for sure I wanted to do it as soon I have a decent version and the documentation website!

**Reply by djowel** [reply]
SO COOL!!! I love the scheme and design!

**Comment by djowel** [comment]
> Now I'm facing a challenge that may be easily fixable from my point of view. Basically I want to cancel a pending callback on the view's timer, so if it was not called before the timeout when cancelled, it is not called at all.

Yes, I see what you mean, and I think it is doable. Would you like to prepare a PR?

**Reply by na2axl** [reply]
Yeah I think I can make it. It all depends on how you want that feature to be. Is it ok to just return the timer? Or do you want a wrapper around it?

**Reply by djowel** [reply]
> Yeah I think I can make it. It all depends on how you want that feature to be. Is it ok to just return the timer? Or do you want a wrapper around it?

I haven't given much thought about it, TBH. But I think I'm going to trust your design instincts. We can always iterate as we know more about the plausible use-cases. Start simple. It's always better to add more features later than to add all the features at once.

**Reply by na2axl** [reply]
@djowel I've just opened a PR here #362

**Comment by djowel** [comment]
This is now implemented. I'd really appreciate it if you can add your app here: please add your library here: https://github.com/cycfi/elements/discussions/352

Cheers!

**Reply by na2axl** [reply]
Just added it!
